llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp

```diff
@@ -1024,6 +1024,15 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
     }
     break;
   }
+  case Intrinsic::amdgcn_wavefrontsize: {
+    // TODO: this is a workaround for the pseudo-generic target one gets with no
+    // specified mcpu, which spoofs its wave size to 64; it should be removed.
+    if ((ST->getCPU().empty() || ST->getCPU() == "generic") &&
+        !ST->getFeatureString().contains("+wavefrontsize"))
```

Comment:
The feature string may also contain a -wavefrontsize.

It's probably safest to ignore the target-features. If we're really going to rely on this target-cpu hack for the library uses, rocm-device-libs is not using an explicit wavefrontsize feature anymore (all the uses were converted to the ballot wave64->wave32 hack)

Reply:
Since this is a temporary hack, what's the point of putting it in the subtarget so that people get ideas, start using it, and then there's even more technical debt? The parsing might have to change anyway once we have a proper generic target (which the current hack is not).

Reply:
That seems reasonable, there's also an argument that the backend likely can't do anything useful without `-mcpu=` set so we could make that an error. The NVIDIA tools basically do that themselves.

Reply:
I see this comment
```
  // Don't crash on invalid devices.
  if (WavefrontSizeLog2 == 0)
    WavefrontSizeLog2 = 5;
```
So might break something.

Reply:
Well, I don't think we should be doing live design of generic (which is part of what got us here anyway), so I'd rather not build even more technical debt around its current form which was meant to be a test only kludge:

```cpp
// The code produced for "generic" is only useful for tests and cannot
// reasonably be expected to execute on any particular target.
```

Which is to say I don't want to what is there now, I want it to not break. I've adjusted the check to cover `generic-hsa` and corrected the commit message.

Reply:
I'd be fine with multiple builds, but right now the AMDGCN infra doesn't support it very well since we'd need to port the ROCm Device Libs to use my build system. Beyond that it'd be pretty easy to just default the triple depending on mcpu and `-mwavefrontsize`.

Reply:
Perhaps we can simply live with `starts_with` for now, as the check won't be infectious and we'll remove it in the future anyway.

Reply:
Unsure if we even need to bother checking for 'generic' since that's not what any of the existing targets use for generic AFAIC. It's just not setting `-mcpu` when you compile, which I think should probably be an error in the backend.